vm/rust/src/lib.rs

```diff
@@ -177,15 +191,83 @@ pub extern "C" fn cairoVMExecute(
             _ => None,
         };
 
-        let txn = Transaction::from_api(sn_api_txn.clone(), contract_class, None);
+        let paid_fee_on_l1: Option<Fee> = match sn_api_txn.clone() {
+            StarknetApiTransaction::L1Handler(_) => {
+                if paid_fees_on_l1.len() == 0 {
+                        report_error(reader_handle, "missing fee paid on l1b".to_string().as_str());
+                        return;
+                }
+                Some(*paid_fees_on_l1.remove(0))
+            }, 
+            _ => None,
+        };
+
+        let txn = Transaction::from_api(sn_api_txn.clone(), contract_class, paid_fee_on_l1);
         if txn.is_err() {
             report_error(reader_handle, txn.unwrap_err().to_string().as_str());
             return;
         }
 
         let res = match txn.unwrap() {
             Transaction::AccountTransaction(t) => t.execute(&mut state, &block_context),
-            Transaction::L1HandlerTransaction(t) => t.execute(&mut state, &block_context),
+            Transaction::L1HandlerTransaction(t) => {
+                // Manually inline L1HandlerTransaction.execute and execute_raw since the
+                // `actual_fee` in Blockifier is zero (i.e., Blockifier doesn't charge for L1
+                // Handler transactions on L2). This way we can get the full
+                // TransactionExecutionResult.
```

Comment:
Do we need to inline this entire thing?

`actual_resources` is a part of `TransactionExecutionInfo`. So we should be able to inline `calculate_tx_fee(&actual_resources, &context.block_context)?;` only?

Reply:
We get `actual_resources` from `calculate_tx_resources`, which requires `calculate_l1_gas_usage`, which requires `execute_call_info`, which requires `t.run_execute`, etc.

Is there a way to inline less? I haven't looked much into the implementation of the functions to see if they modify their inputs, of some of them take mutable references so I played it safe.

Reply:
Correct me if I am wrong but I believe this works too

``` rust
        let res = match txn.unwrap() {
            Transaction::AccountTransaction(t) => t.execute(&mut state, &block_context),
            Transaction::L1HandlerTransaction(t) => {
                let maybe_execution_info = t.execute(&mut state, &block_context);
                if maybe_execution_info.is_err() {
                    maybe_execution_info
                } else {
                    let mut execution_info = maybe_execution_info.unwrap();
                    execution_info.actual_fee = calculate_tx_fee(&execution_info.actual_resources, &block_context).unwrap();
                    Ok(execution_info)    
                }
            },
        };
```